### PR TITLE
feat(InteractionCollector): reworked to be more generic

### PR DESCRIPTION
**Pull request by monbrey**
Another half-baked idea that comes out of late night conversations and early morning coding.

This refactors the `MessageComponentInteractionCollector` into a generic `InteractionCollector` that can be used to collect any sort of interaction (yes, even Slash Commands).

It also removes the need for it to be bound to any particular message or channel, though it still can be should you want to do that. These are still separate options to the filter as it allows the Collector to also listen to the relevant DELETE events and cancel itself accordingly. It will use these based on the most specific option first, so Message > Channel > Guild.

Draft for now - I'd still like reviews and opinions from other people. I intend to add some checking to the constructor to account for options that don't make sense together, e.g. setting a `componentType` option if `interactionType` is set for slash commands.

**Status and versioning classification**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: src/structures/InteractionCollector.js [hunk 1]
```diff
@@ -17,26 +23,49 @@ const { Events } = require('../util/Constants');
...
      * @type {?Message}
      */
-    this.message = source instanceof require('./Message') ? source : null;
+    this.message = options.message ?? null;
 
     /**
-     * The source channel from which to collect message component interactions
-     * @type {TextChannel|DMChannel|NewsChannel}
+     * The channel from which to collect interactions, if provided
+     * @type {?TextChannel|DMChannel|NewsChannel}
```
Comment (almeidx) [inline src/structures/InteractionCollector.js:42]:
```suggestion
     * @type {?(TextChannel|DMChannel|NewsChannel)}
```

**Review by almeidx — CHANGES_REQUESTED** [review]

File: src/structures/InteractionCollector.js (hunk 1 above)
Comment (NotSugden) [reply src/structures/InteractionCollector.js:42]: isn't it the same type either way

Comment (AustrianDev) [reply src/structures/InteractionCollector.js:42]: Doesn't TextBasedChannel work?

Comment (monbrey) [reply src/structures/InteractionCollector.js:42]: TextBasedChannel isn't a jsdoc type, its an interface applied to other classes.

File: typings/index.d.ts [hunk 2]
```diff
@@ -1183,6 +1183,34 @@ declare module 'discord.js' {
...
 
+  export class InteractionCollector<T extends Interaction> extends Collector<Snowflake, T> {
+    constructor(client: Client, options?: InteractionCollectorOptions<T>);
+    private _handleMessageDeletion(message: Message): void;
+    private _handleChannelDeletion(channel: GuildChannel): void;
+    private _handleGuildDeletion(guild: Guild): void;
+
+    public channel: TextChannel | NewsChannel | DMChannel | null;
+    public componentType: MessageComponentType | null;
+    public empty(): void;
```
Comment (SpaceEEC) [inline typings/index.d.ts:1194]: Must have slipped through before, but methods should go below properties.

**Review by SpaceEEC — DISMISSED** [review]

File: src/structures/InteractionCollector.js [hunk 3]
```diff
@@ -51,23 +80,26 @@ class MessageComponentInteractionCollector extends Collector {
...
+      this.client.on(Events.MESSAGE_DELETE, this._handleMessageDeletion);
+    }
+    if (this.channel) {
+      this._handleChannelDeletion = this._handleChannelDeletion.bind(this);
+      this.client.on(Events.CHANNEL_DELETE, this._handleChannelDeletion);
+    }
+    if (this.guild) {
+      this._handleGuildDeletion = this._handleGuildDeletion.bind(this);
+      this.client.on(Events.GUILD_DELETE, this._handleGuildDeletion);
+    }
```
Comment (vladfrangu) [inline src/structures/InteractionCollector.js:95]: Small looks nit, alongside a request to increment the max listeners per client.on call :D

```suggestion
    if (this.message) {
      this._handleMessageDeletion = this._handleMessageDeletion.bind(this);
      this.client.on(Events.MESSAGE_DELETE, this._handleMessageDeletion);
    }

    if (this.channel) {
      this._handleChannelDeletion = this._handleChannelDeletion.bind(this);
      this.client.on(Events.CHANNEL_DELETE, this._handleChannelDeletion);
    }

    if (this.guild) {
      this._handleGuildDeletion = this._handleGuildDeletion.bind(this);
      this.client.on(Events.GUILD_DELETE, this._handleGuildDeletion);
    }
```

**Review by vladfrangu — DISMISSED** [review]

File: typings/index.d.ts [hunk 4]
```diff
@@ -1183,6 +1183,34 @@ declare module 'discord.js' {
...
+    public guild: Guild | null;
+    public interactionType: InteractionType | null;
+    public message: Message | null;
+    public options: InteractionCollectorOptions<T>;
+    public total: number;
+    public users: Collection<Snowflake, User>;
+
+    public collect(interaction: Interaction): Snowflake;
+    public dispose(interaction: Interaction): Snowflake;
+    public on(event: 'collect' | 'dispose', listener: (interaction: T) => Awaited<void>): this;
```
Comment (NotSugden) [inline typings/index.d.ts:1205]:
```suggestion
    public on(event: 'collect' | 'dispose', listener: (interaction: T) => void): this;
```
void works just the same, no real reason to use `Awaited<void>` over just `void` ?

Comment (DTrombett) [reply typings/index.d.ts:1205]: It's the same as `Client#on()`, which uses Awaited too

Comment (monbrey) [reply typings/index.d.ts:1205]: Pretty sure the reason it was switched to `Awaited<void>` is because otherwise you can't have async listeners.

Comment (DTrombett) [reply typings/index.d.ts:1205]: Yeah just found #5602

Comment (NotSugden) [reply typings/index.d.ts:1205]: didn't get any errors on [ts playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZgVzMGBeGAKOYBcmAeeARiCADYCmAhmAJRoB8MAbiAJYAm9qT2GUAJwQVaAbgBQ4xMgxUIATxncmAb3EwNMARSgIBYGACJoAtmADmhiQF8xk6cAwZlMNZq069BsBQDuMAAUBEABbNggKAB4TM3MGDG0IcmYKRg8kshSMY0FYw1o7W1EgA) 🤷 strange

Comment (almeidx) [reply typings/index.d.ts:1205]: > didn't get any errors on [ts playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAZgVzMAXDAFHMb0A80BGIIANgKYCGYAlDALwB8MAbiAJYAmtjL7H9GLDnwwipSjXpNWnbkyzooAJwRlqAbgBQmxMnQUIATz1yYAb00wrMJWSgIlYGACJoStmADmzrQF8N2rrA6OimFtY2dg5OYGQA7jAACkogALZsEGQAPG4engzothCkzGRSkcUkpeiuynnO1AH+6kA)  strange

From the pull request @DTrombett linked
> When using `client.on` with a callback that returns a `Promise`, eslint will then complain if the `no-misused-promises` (from `@typescript-eslint` and various plugins) is enabled, as `Promise<?>` does not match `void`.

Comment (NotSugden) [reply typings/index.d.ts:1205]: ah, makes sense

File: src/structures/InteractionCollector.js (hunk 3 above)
Comment (monbrey) [reply src/structures/InteractionCollector.js:95]: I don't think it's necessary to do it per call? The maxListeners property determines how many listeners can be bound to each event, not in total. This is only binding one listener to each event, so incrementing once should be sufficient.

I'll move the increment call to the top though.

**Review by kyranet — APPROVED** [review]

**Comment by iCrawl** [comment]
This needs a rebase.

**Review by SpaceEEC — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Pull request closed without merging**